packages/workflow/src/internals.ts

```diff
@@ -534,14 +542,116 @@ export class Activator implements ActivationHandler {
+    }
+    if (!name) {
+      throw new TypeError('Missing activation update name');
+    }
+    if (!this.updateHandlers.has(name)) {
+      // TODO (dan): Signal is able to handle this situation more gracefully by
+      // using this.bufferedSignals. Should something analogous exist for
+      // Update?
+      this.rejectUpdate(updateId, ApplicationFailure.nonRetryable(`Update has no handler: ${name}`));
+      return;
+    }
+    const execute = composeInterceptors(this.interceptors.inbound, 'handleUpdate', this.updateNextHandler.bind(this));
+
+    const validate = composeInterceptors(
+      this.interceptors.inbound,
+      'validateUpdate',
+      this.validateUpdateNextHandler.bind(this)
+    );
+
+    const makeInput = (): UpdateInput => ({
+      updateId,
+      args: arrayFromPayloads(this.payloadConverter, activation.input),
+      name,
+      headers: headers ?? {},
+    });
+
+    // The implementation below is responsible for upholding, and constrained
+    // by, the following contract:
+    //
+    // 1. During validation, any error must fail the Update; during the Update
+    //    itself, Temporal errors fail the Update whereas other errors fail the
+    //    activation.
+    //
+    // 2. The handler must not see any mutations of the arguments made by the
+    //    validator.
+    //
+    // 3. Any error when decoding/deserializing input must be caught and result
+    //    in rejection of the Update before it is accepted, even if there is no
+    //    validator.
+    //
+    // 4. The initial synchronous portion of the (async) Update handler should
+    //    be executed after the (sync) validator completes such that there is
+    //    minimal opportunity for a different concurrent task to be scheduled
+    //    between them.
+    //
+    // 5. The stack trace view provided in the Temporal UI must not be polluted
+    //    by promises that do not derive from user code. This implies that
+    //    async/await syntax may not be used.
+    //
+    // Note that there is a deliberately unhandled promise rejection below.
+    // These are caught elsewhere and fail the corresponding activation.
+    let input: UpdateInput;
+    try {
+      if (runValidator) {
+        validate(makeInput());
+      }
+      input = makeInput();
+    } catch (error) {
+      this.rejectUpdate(updateId, error);
+      return;
+    }
+    this.acceptUpdate(updateId);
+    untrackPromise(
+      execute(input)
+        .then((result) => this.completeUpdate(updateId, result))
+        .catch((error) => {
+          if (error instanceof TemporalFailure) {
+            this.rejectUpdate(updateId, error);
+          } else {
+            throw error;
+          }
+        })
+    );
+  }
+
+  // Intentionally non-async function so this handler doesn't show up in the
+  // stack trace. The user's handler may be a normal function or an async
+  // function.
+  protected updateNextHandler({ name, args }: UpdateInput): Promise<unknown> {
```

Comment:
This one should be async.

Sorry, this knowledge that's only in my head. I need to formalize it somewhere.

Reply:
I'd like to understand this before changing it.

Reply:
The idea is that if it's an async function it becomes part of the stack trace and we can cut off the stack in the right place. We have a utility that looks up `.*NextHandler` methods to perform the cutoff.

Reply:
OK, so update and signal version async, and query version non-async (but returning promise).